### PR TITLE
Add caching support for Lychee with --cache flag detection

### LINKED ISSUE #291: Automatically set up cache
Given that this is a composite action, should we automatically set up a cache for it? It is effectively a no-op if users don't pass `--cache` but will benefits users that want to have one because they don't need to set it up manually then?

### PR DESCRIPTION
**Pull request by mre**
This adds GitHub Actions caching for the `.lycheecache` file when the `--cache` flag is present.

## Important Notes

1. The cache is only for the `.lycheecache` file, not the lychee binary
2. Caching is only enabled when `--cache` is present in the lychee arguments
3. The cache is saved even if the lychee run fails (using `!cancelled()` condition)
4. The cache key uses `github.sha` for uniqueness with fallback to previous caches
5. The cache path respects the `workingDirectory` input

## Usage Example

When users include `--cache` in their arguments:

```yaml
- uses: lycheeverse/lychee-action@main
  with:
    args: "--cache --verbose --no-progress './**/*.md'"
```

The action will:
1. Detect the `--cache` flag
2. Restore any existing `.lycheecache` file before running
3. Save the updated `.lycheecache` file after running

Without `--cache`, no GitHub Actions caching occurs, preventing any issues in restricted environments.

Fixes #291.

### CODE REVIEW

**Comment by mre** [comment]
The question is what will happen with existing workflows that might have a manual cache setup. 
I think the behavior depends a bit on the setup:

- If the cache key is identical, it might be a no-op? (I hope so at least)
- If the cache key is different, we store/restore the cache twice, which might be a problem.

(Correct me if I'm wrong here.)

I wonder how to detect if there's an existing cache setup. There probably is no way.
The easiest "fix" I can think of is to mention it in the release notes and release the change under a new major version v3.

**Comment by Arteiii** [comment]
One thought: using a dedicated action input like `cache: true` (defaulting to false) might be cleaner than checking for `--cache` in the args. That way, it avoids changing the behavior of `--cache` and keeps things more explicit for users with custom caching setups

**Comment by thomaseizinger** [comment]
Thank you for working on this!

> One thought: using a dedicated action input like `cache: true` (defaulting to false) might be cleaner than checking for `--cache` in the args. That way, it avoids changing the behavior of `--cache` and keeps things more explicit for users with custom caching setups

As someone setting the action up from scratch, you might be wondering then "How many times do I need to tell this action that it should use a cache? :)"

In my experience, many actions enable a cache by default these days because it is basically always what you want and the action maintainers likely know best how the cache should be managed.

My vote would go down for a new major version of the action, with an on-by-default cache and perhaps with `--cache` also being passed by default.

**Comment by Arteiii** [comment]
Yeee, this works as well and is probably clearer and easier for people using the action. I actually thought about it too, but wasn’t sure if it was worth pushing another release just for this change, especially since there aren't any other updates lined up right now.

**Comment by mre** [comment]
Fair point. I think we could wait a bit before we merge this. People interested could already use it (and remove the existing cache if any):

```yaml
- name: Link Checker
  id: lychee
  uses: lycheeverse/lychee-action@cache
  with:
    fail: false
```

In fact, I believe it would be best if someone could test it for a while to fix any edge-cases. @thomaseizinger wanna volunteer? 😬

**Comment by thomaseizinger** [comment]
> @thomaseizinger wanna volunteer? 😬

Definitely! We are facing 429s on our daily cron suddenly and I am planning to see if the cache mitigates that.

**Comment by mre** [comment]
Maybe another case of https://github.com/lycheeverse/lychee-action/issues/289?

**Comment by thomaseizinger** [comment]
Is this branch safe to depend on? We are hash-pinning all actions in our CI so if you'd like me to test this, I'd ask to please on force-push to this branch :)

**Comment by mre** [comment]
To be on the safe side, you can also use the commit sha:

```
lycheeverse/lychee-action@e203314714efe5700a013ace5248a697313082b2
```

This won't change.

**Comment by Arteiii** [comment]
still think adding an explicit action argument (`cache: true`) would make things cleaner (see: https://github.com/lycheeverse/lychee-action/pull/292#issuecomment-2869868825).

It could default to true, but still check for --cache to avoid conflicts.

That said, --cache shouldn't automatically enable GitHub caching

keeping that opt-in gives users more flexibility for custom setups.

Otherwise, looks good to me!

**Comment by Arteiii** [comment]
I can't think of a setup where the built-in cache wouldn't be enough, but custom setups might exist. To avoid breaking anything, I’d suggest:

- If --cache is set alone, just use default Lychee behavior.

- If --cache (or a custom location) and cache: true in the config are set, raise an error to avoid confusion.

This keeps things explicit without changing behavior for existing setups.

**Comment by Arteiii** [comment]
> - If --cache is set alone, just use default Lychee behavior.

this also makes things more explicit, since we shouldn't modify the Action's behavior based on arguments passed to Lychee via `args: ....`

**Comment by thomas-zahner** [comment]
In general im in favour of the proposal of @mre since it's just easier for the vast majority of users. In my opinion it boils down to the following: is it still possible to manually override the caching behaviour introduced in this PR?

If so, I think we should go with the approach suggested by @mre. When bumping the major version of the action we are allowed to introduce breaking changes. Users with custom caching logic could stay on the current version or adapt and update.

File: action.yml [hunk 1]
```diff
@@ -102,6 +116,16 @@ runs:
         # Install lychee from the temporary directory
         install -t "$HOME/.local/bin" -D "${{ steps.lychee-setup.outputs.temp_dir }}/lychee"
       shell: bash
+
+    - name: Restore lychee cache
+      id: restore-cache
+      if: steps.cache-check.outputs.cache_enabled == 'true'
+      uses: actions/cache/restore@v4
```
Comment (jonatan-ivanov) [inline action.yml]: fyi: v5 was recently released

**Comment by Arteiii** [comment]
> In general im in favour of the proposal of @mre since it's just easier for the vast majority of users. In my opinion it boils down to the following: is it still possible to manually override the caching behaviour introduced in this PR?
> 
> If so, I think we should go with the approach suggested by @mre. When bumping the major version of the action we are allowed to introduce breaking changes. Users with custom caching logic could stay on the current version or adapt and update.

I'm fully in favor of the proposal to add dedicated `cache` and `cache-path` inputs. Since we are planning a major release, this is the right time to improve the UX without worrying about breaking changes for existing users.

To be clear: adding these fields doesn't take anything away from power users. The `args` field remains available for those who need custom configurations. This is purely an additive, "opt-in" simplification for the 90% of users who just want caching to work without digging into CLI flags.

If there are specific technical downsides or maintenance concerns that haven't been mentioned yet, please bring them up now so we can address them. Otherwise, the benefits of making the action more intuitive for the general community seem to far outweigh any perceived "abstraction" costs.

File: action.yml (hunk 1 above)
Comment (Arteiii) [inline action.yml]:
```suggestion
      uses: actions/cache/restore@v5
```

**Review by Arteiii — CHANGES_REQUESTED** [review]
https://github.com/lycheeverse/lychee-action/pull/292#issuecomment-4382458885

As I previously stated, I remain to be opposed to this course of action

**Comment by mre** [comment]
Updated this PR to take the explicit-input route discussed above.

What changed:

- Added `cache: true` to opt into GitHub Actions cache integration.
- Added `cachePath`, relative to `workingDirectory`, for users who want the cache file somewhere other than `.lycheecache`.
- No longer enables GitHub Actions caching just because `args` contains `--cache`.
- When `cache: true` is set, the action passes lychee `--cache` itself.
- If users set both `cache: true` and `--cache` / `--cache-path` in `args`, the action now fails early with a clear error.
- Updated docs and added a CI regression test for the ambiguous double-cache configuration.
- Merged current `master`, so the PR is no longer conflicting.

One implementation detail: current lychee has `--cache` but no `--cache-path` CLI flag, so `cachePath` is handled by the action. For non-default paths, the action restores/saves that path and symlinks `.lycheecache` to it before running lychee.

**Comment by mre** [comment]
@Arteiii I think the current branch now follows your suggested direction:

- `cache: true` is explicit and defaults to false.
- Passing `--cache` in `args` no longer enables GitHub Actions caching.
- `cache: true` plus `--cache` / `--cache-path` in `args` fails explicitly.
- `cachePath` covers custom cache locations.
- Cache actions are now on `v6`.

Could you take another look when you have a minute?

File: README.md [hunk 2]
```diff
@@ -97,34 +99,36 @@ default `GITHUB_TOKEN`, you can create a [personal access
...
 
+This turns on lychee's request cache and stores the cache file with GitHub Actions cache. By default, the cache lives at `.lycheecache` relative to `workingDirectory`. Use `cachePath` to choose a different relative path:
+
+```yml
 - name: Run lychee
   uses: lycheeverse/lychee-action@v2
   with:
-    args: "--root-dir "$(pwd)" --cache --max-cache-age 1d ."
+    cache: true
+    cachePath: website/.lycheecache
```
Comment (Arteiii) [inline README.md:121]: For the standard use case the cache path shouldn't be required or am I wrong here?

**Review by Arteiii — COMMENTED** [review]
Besides the readme lgtm
